src/compiler/scala/tools/nsc/backend/jvm/ClassHandler.scala

```diff
@@ -0,0 +1,298 @@
+          }
+      }
+    }
+    def tryStealing:Option[Runnable]
+  }
+  private final class SyncWritingClassHandler(val unitInfoLookup: UnitInfoLookup,
+                val postProcessor: PostProcessor, val cfWriter: ClassfileWriter)
+    extends WritingClassHandler((r) => r.run()) {
+
+    override def toString: String = s"SyncWriting [$cfWriter]"
+
+    override def tryStealing: Option[Runnable] = None
+  }
+  private final case class ExecutorServiceInfo( maxThreads:Int, javaExecutor : ExecutorService, queue:BlockingQueue[Runnable])
+
+  private final class AsyncWritingClassHandler(val unitInfoLookup: UnitInfoLookup,
+                                               val postProcessor: PostProcessor,
+                                               val cfWriter: ClassfileWriter,
+                                               val executorServiceInfo: ExecutorServiceInfo)
+    extends WritingClassHandler(executorServiceInfo.javaExecutor) {
+
+    override def toString: String = s"AsyncWriting[additional threads:${executorServiceInfo.maxThreads} writer:$cfWriter]"
+
+    override def close(): Unit = {
+      super.close()
+      executorServiceInfo.javaExecutor.shutdownNow()
+    }
+
+    override def tryStealing: Option[Runnable] = Option(executorServiceInfo.queue.poll())
+  }
+
+}
+//we avoid the lock on frontendSync for the common case, when compiling to a single target
+sealed trait UnitInfoLookup {
+  def outputDir(source:AbstractFile) : AbstractFile
+  val frontendAccess: PostProcessorFrontendAccess
+}
+final class SingleUnitInfo(val frontendAccess: PostProcessorFrontendAccess, constantOutputDir:AbstractFile) extends UnitInfoLookup {
+  override def outputDir(source: AbstractFile) = constantOutputDir
+}
+final class LookupUnitInfo(val frontendAccess: PostProcessorFrontendAccess) extends UnitInfoLookup {
+  lazy val outputDirectories = frontendAccess.compilerSettings.outputDirectories
+  override def outputDir(source: AbstractFile) = outputDirectories.outputDirFor(source)
+}
+sealed trait SourceUnit {
+  def withBufferedReporter[T](fn: => T): T
+
+  val outputDir: AbstractFile
+  val outputPath: java.nio.file.Path
+  def sourceFile:AbstractFile
+
+}
+final class UnitResult(unitInfoLookup: UnitInfoLookup, classes_ : List[GeneratedClass], val source:SourceFile) extends SourceUnit with BackendReporting {
+  lazy val outputDir = unitInfoLookup.outputDir(source.file)
+
+  override def sourceFile = source.file
+
+  lazy val outputPath = outputDir.file.toPath
+  private var classes: List[GeneratedClass] = classes_
+
+  def copyClasses = classes
+
+  def takeClasses(): List[GeneratedClass] = {
+    val c = classes
+    classes = Nil
+    c
+  }
+
+  /** the main async task submitted onto the scheduler */
+  var task: Future[Unit] = _
+
+  /** the final completion which may occur after the task completes
+    * this allows the use of async completions */
```

Comment:
I don't understand the what this promise is for.

I also don't understand the comment. As far as I can see, the main `task` Future sets the `result` promise before it completes, so the `result` will complete before (not after) the `task`.

Reply:
probably a hangover from the preceeding versionthat had async file writing - this allowed for the classHandler to complete, but he  IOs hadn't - will check

Reply:
removed now